admindb/sqlite/invites.go

```diff
@@ -0,0 +1,206 @@
+package sqlite
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"database/sql"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/friendsofgo/errors"
+	"github.com/mattn/go-sqlite3"
+	"github.com/volatiletech/null/v8"
+	"github.com/volatiletech/sqlboiler/v4/boil"
+	"github.com/volatiletech/sqlboiler/v4/queries/qm"
+
+	"github.com/ssb-ngi-pointer/go-ssb-room/admindb"
+	"github.com/ssb-ngi-pointer/go-ssb-room/admindb/sqlite/models"
+	refs "go.mindeco.de/ssb-refs"
+)
+
+// make sure to implement interfaces correctly
+var _ admindb.InviteService = (*Invites)(nil)
+
+// Invites implements the admindb.InviteService.
+// Tokens are stored as sha256 hashes on disk to protect against attackers gaining database read-access.
+type Invites struct {
+	db *sql.DB
+
+	allowList *AllowList
+}
+
+const tokenLength = 50
+
+// Create creates a new invite for a new member. It returns the token or an error.
+// createdBy is user ID of the admin or moderator who created it.
+// aliasSuggestion is optional (empty string is fine) but can be used to disambiguate open invites. (See https://github.com/ssb-ngi-pointer/rooms2/issues/21)
+// The returned token is base64 URL encoded and has tokenLength when decoded.
+func (i Invites) Create(ctx context.Context, createdBy int64, aliasSuggestion string) (string, error) {
+	isValidSuggestion := aliasSuggestion != ""
+	var newInvite = models.Invite{
+		CreatedBy:       createdBy,
+		AliasSuggestion: null.NewString(aliasSuggestion, isValidSuggestion),
+	}
+
+	tokenBytes := make([]byte, tokenLength)
+
+	err := transact(i.db, func(tx *sql.Tx) error {
+
+		inserted := false
+	trying:
+		for tries := 100; tries > 0; tries-- {
+			// generate an invite code
+			rand.Read(tokenBytes)
+
+			// hash the binary of the token for storage
+			h := sha256.New()
+			h.Write(tokenBytes)
+			newInvite.Token = fmt.Sprintf("%x", h.Sum(nil))
+
+			// insert the new invite
+			err := newInvite.Insert(ctx, tx, boil.Infer())
+			if err != nil {
+				var sqlErr sqlite3.Error
+				if errors.As(err, &sqlErr) && sqlErr.ExtendedCode == sqlite3.ErrConstraintUnique {
+					// generated an existing token, retry
+					continue trying
+				}
+				return err
+			}
+			inserted = true
+			break // no error means it worked!
+		}
+
+		if !inserted {
+			return errors.New("admindb: failed to generate an invite token in a reasonable amount of time")
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return "", err
+	}
+
+	return base64.URLEncoding.EncodeToString(tokenBytes), nil
+}
+
+// Consume checks if the passed token is still valid. If it is it adds newMember to the members of the room and invalidates the token.
+// If the token isn't valid, it returns an error.
+// Tokens need to be base64 URL encoded and when decoded be of tokenLength.
+func (i Invites) Consume(ctx context.Context, token string, newMember refs.FeedRef) (admindb.Invite, error) {
```

Comment:
is it super faux-pas to do like `func (this Invites) ... <function>`? `i` is so ingrained for me as an iteration variable / useless int haha, had to do a double-take when reading the code & skimming the function signature. kinda on the nitpicky side tho

(or `func (invites Invites)` but borders on verbose)

Reply:
also i recently read https://macwright.com/2021/02/17/the-naming-of-things.html so i think that's definitely a lens i am reviewing through ^^

Reply:
hm yea.. i don't know why but i really don't like `this` or `self` for these. Definitely a style preference thing.

I was starting out with `db` but since in most cases here the only thing in the receiver struct is the database handle `db`, I ended up with a lot of `db.db` which also felt silly.